R/functions-Chromatograms.R

```diff
@@ -121,4 +141,38 @@ Chromatograms <- function(data, phenoData, ...) {
     }
 }
 
-
+#' Helper function to extract mz, precursorMz or productMz from a Chromatograms
+#' object
+#'
+#' @author Johannes Rainer
+#'
+#' @noRd
+.mz_chromatograms <- function(x, mz = "mz") {
+    mz <- match.arg(mz, c("mz", "precursorMz", "productMz"))
+    if (!nrow(x))
+        return(matrix(nrow = 0, ncol = 2, dimnames = list(character(),
+                                                          c("mzmin", "mzmax"))))
+    ## If we've got the values in the featureData, use these.
+    if (mz %in% c("precursorMz", "productMz"))
+        vl <- rep(paste0(sub(mz, pattern = "Mz", replacement = ""),
+                         "IsolationWindowTargetMZ"), 2)
+    else
+        vl <- c("mzmin", "mzmax")
+    if (all(vl %in% fvarLabels(x))) {
+        ## Want to return a matrix, not a data.frame
+        cbind(mzmin = fData(x)[, vl[1]], mzmax = fData(x)[, vl[2]])
```

Comment:
I'll check - I think there was something behind using `cbind` here (performance wise).

Reply:
`cbind` is slightly faster than `as.matrix`. Not that it matters here, I just prefer using `cbind` and extract individual columns from a `data.frame` instead of anything that involves accessing multiple columns at once in a `data.frame`, as that might/can involve copying of the data, while accessing single columns of a `data.frame` never copies the data.

Reply:
Oh, cool. I wasn't aware of that. You are right:
```r
library("microbenchmark")
set.seed(2017)
n <- 1e5
d <- data.frame(a=sample(n), b=sample(n), c=sample(n))

f1 <- function(x, vl=c("a", "b"))cbind(mzmin=x[, vl[1L]], mzmax=x[, vl[2L]])
f2 <- function(x, vl=c("a", "b"))matrix(c(x[, vl[1L]], x[, vl[2L]]), ncol=2L,
                                        dimnames=list(NULL, c("mzmin", "mzmax")))

all.equal(f1(d), f2(d))
# [1] TRUE

microbenchmark(f1(d), f2(d))
# Unit: microseconds
#   expr     min       lq     mean   median      uq      max neval
#  f1(d) 112.814 122.4080 203.0495 126.7465 139.109 3693.441   100
#  f2(d) 505.302 523.0355 638.2103 526.0990 536.168 4133.861   100
```